subsys/net/lib/lwm2m/lwm2m_rw_cbor.c

```diff
@@ -0,0 +1,554 @@
+		LOG_WRN("unable to decode a 32-bit integer value");
+		return -EBADMSG;
+	}
+
+	int len = ICTX_CBOR_R_SZ(states[0].payload, in);
+
+	in->offset += len;
+
+	return len;
+}
+
+static int get_float(struct lwm2m_input_context *in, double *value)
+{
+	ZCBOR_STATE_D(states, 0, ICTX_BUF_R_PTR(in), ICTX_BUF_R_LEFT_SZ(in),  1);
+
+	if (!zcbor_float32_decode(states, (float *)value)) {
+		LOG_INF("unable to decode a 32-bit floating-point value - trying width 64");
+		if (!zcbor_float64_decode(states, value)) {
+			LOG_ERR("unable to decode a floating-point value");
+			return -EBADMSG;
+		}
+	}
+
+	int len = ICTX_CBOR_R_SZ(states[0].payload, in);
+
+	in->offset += len;
+
+	return len;
+}
+
+static int get_string(struct lwm2m_input_context *in, uint8_t *value, size_t buflen)
+{
+	struct zcbor_string hndl;
+	int len;
+
+	ZCBOR_STATE_D(states, 0, ICTX_BUF_R_PTR(in), ICTX_BUF_R_LEFT_SZ(in),  1);
+
+	if (!zcbor_tstr_decode(states, &hndl)) {
+		LOG_WRN("unable to decode a string");
+		return -EBADMSG;
+	}
+
+	len = MIN(buflen-1, hndl.len);
+
+	memcpy(value, hndl.value, len);
+
+	value[len] = '\0';
+
+	len = ICTX_CBOR_R_SZ(states[0].payload, in);
+
+	in->offset += len;
+
+	return len;
+}
+
+/* Gets time decoded as a numeric string.
+ *
+ * return 0 on success, -EBADMSG if decoding fails or -EFAIL if value is invalid
+ */
+static int get_time_string(struct lwm2m_input_context *in, int64_t *value)
+{
+	char time_str[sizeof("4294967295")] = { 0 };
+	struct zcbor_string hndl = { .value = time_str, .len = sizeof(time_str) - 1 };
+
+	ZCBOR_STATE_D(states, 0, ICTX_BUF_R_PTR(in), ICTX_BUF_R_LEFT_SZ(in),  1);
+
+	if (!zcbor_tstr_decode(states, &hndl)) {
+		return -EBADMSG;
+	}
+
+	/* TODO: decode date/time string */
+	LOG_DBG("decoding a date/time string not supported");
+
+	return -ENOTSUP;
+}
+
+/* Gets time decoded as a numerical.
+ *
+ * return 0 on success, -EBADMSG if decoding fails
+ */
+static int get_time_numerical(struct lwm2m_input_context *in, int64_t *value)
+{
+	ZCBOR_STATE_D(states, 0, ICTX_BUF_R_PTR(in), ICTX_BUF_R_LEFT_SZ(in),  1);
+
+	if (!zcbor_int64_decode(states, value)) {
+		LOG_WRN("unable to decode seconds since Epoch");
+		return -EBADMSG;
+	}
+
+	return 0;
+}
+
+static int get_time(struct lwm2m_input_context *in, int64_t *value)
+{
+	uint32_t tag;
+	int tag_sz = 0;
+	int data_len;
+	int ret;
+	bool success;
+
+	ZCBOR_STATE_D(states, 0, ICTX_BUF_R_PTR(in), ICTX_BUF_R_LEFT_SZ(in),  1);
+
+	success = zcbor_tag_decode(states, &tag);
+
+	if (success) {
+		tag_sz = ICTX_CBOR_R_SZ(states[0].payload, in);
+		in->offset += tag_sz;
+
+		switch (tag) {
+		case ZCBOR_TAG_TIME_NUM:
+			ret = get_time_numerical(in, value);
+			if (ret < 0) {
+				goto error;
+			}
+			break;
+		case ZCBOR_TAG_TIME_TSTR:
+			ret = get_time_string(in, value);
+			if (ret < 0) {
+				goto error;
+			}
+			break;
+		default:
+			LOG_WRN("expected tagged date/time, got tag %" PRIu32 "", tag);
+			return -EBADMSG;
+		}
+	} else { /* TODO: Tags are optional, right? */
```

Comment:
What about this TODO, it's more of a question than the actual work to do, could be misleading for people reading the code (like me).

Reply:
Understandable, fixed.